### PR TITLE
MULE-11848: Replace direct buffers with pooled heap buffers

### PR DESCRIPTION
**Pull request by marianogonzalez**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/AbstractCursorStreamProviderAdapter.java [hunk 1]
```diff
@@ -63,5 +67,12 @@ public Event getCreatorEvent() {
     return creatorEvent;
   }
 
+  /**
+   * @return the {@link ByteBufferManager} that <b>has</b> to be used to allocate byte buffers
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/AbstractCursorStreamProviderAdapter.java]: has to -> `must` to put more enphasis?

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/AbstractInputStreamBuffer.java [hunk 2]
```diff
@@ -117,6 +124,13 @@ protected static ReadableByteChannel openStreamChannel(InputStream stream) {
   protected abstract boolean canBeExpanded();
 
   /**
+   * @return the {@link ByteBufferManager} that <b>has</b> to be used to allocate byte buffers
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/AbstractInputStreamBuffer.java]: has to -> `must` to put more enphasis?

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/FileStoreInputStreamBuffer.java [hunk 3]
```diff
@@ -209,9 +199,10 @@ private File createTempFile(String prefix, String suffix) throws RuntimeExceptio
     }
 
     if (!TEMP_DIR.exists()) {
-      throw new RuntimeException(format("Temp directory '%s' does not exits. Please check the value of the '%s' system property.",
-                                        TEMP_DIR.getAbsolutePath(),
-                                        TEMP_DIR_SYSTEM_PROPERTY));
+      throw new MuleRuntimeException(
+                                     createStaticMessage(format("Temp directory '%s' does not exits. Please check the value of the '%s' system property.",
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/FileStoreInputStreamBuffer.java]: exits -> exist

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+          notification.getValue().close();
+        } catch (Exception e) {
+          if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Found exception trying to dispose buffer pool for capacity " + notification.getKey(), e);
+          }
+        }
+      }).build(new CacheLoader<Integer, ObjectPool<ByteBuffer>>() {
+
+        @Override
+        public ObjectPool<ByteBuffer> load(Integer capacity) throws Exception {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java]: lambda?

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  @Override
+  public void dispose() {
+    pools.invalidateAll();
+  }
+
+  private ObjectPool<ByteBuffer> createPool(int capacity) {
+    GenericObjectPool.Config config = new GenericObjectPool.Config();
+    config.maxIdle = DEFAULT_MAX_IDLE;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java]: make this values configurable (add TODO)

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/factory/AbstractCursorStreamProviderFactory.java [hunk 6]
```diff
@@ -57,7 +62,15 @@
   }
 
   /**
+   * @return the {@link ByteBufferManager} that <b>has</b> to be used to allocate byte buffers
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/factory/AbstractCursorStreamProviderFactory.java]: has to -> `must` to put more enphasis?

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java (hunk 5 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java]: not a functional interface

Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java]: What's the purpose of doing that? This class is not API and serves one and only one purpose. If I make it configurable, there will be nobody configuring it. There's no point in adding that until a use case appears

File: core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java (hunk 4 above)
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java]: not a functional interface

**Pull request merged**